Freebsd futex

LorrensP-2158466 opened the pull request:
Fixes #3553.

Currently This implements the WAIT and WAKE operations of the `_umtx_op` syscall.

I have not yet changed `ci/ci.sh` because I don't exactly know which filters to pass. Currently if I test it locally only the the tests that use `libc::cpuset_getaffinity` fail.

LorrensP-2158466 commented:
@rustbot ready

RalfJung commented:
Thanks for the PR! We have a significant backlog currently so unfortunately it could take a while until you get a proper review, sorry.

> I have not yet changed ci/ci.sh because I don't exactly know which filters to pass. Currently if I test it locally only the the tests that use libc::cpuset_getaffinity fail.

You should add `sync` to have it test the libstd synchronization primitives.

Also, please add a test that directly calls the FreeBSD functions.

saethlin commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,219 @@
+//! Contains FreeBSD-specific synchronization functions
+
+use core::time::Duration;
+use std::u64;
+
+use rustc_span::sym::TryFrom;
```
```suggestion
```

Are these IDE-generated suggestions? You shouldn't need either of them, and `TryFrom` (if it's even accessible from this path) is an accidental re-export.

LorrensP-2158466 replied:
Oops you are right :) ! I noticed them but still forgot to remove them.

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+use crate::concurrency::sync::FutexRef;
+use crate::*;
+
+pub struct FreeBSDFutex {
```
Please follow standard Rust naming conventions and call this `FreeBsdFutex`.

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+        let obj = this.read_pointer(obj)?;
+        let op = this.read_scalar(op)?.to_i32()?;
+        let val = this.read_target_usize(val)?;
+        let uaddr2 = this.read_pointer(uaddr2)?;
```
Why is `uaddr` being skipped here?

LorrensP-2158466 replied:
Because it can be treated as a pointer to `timespec` or as a value equalling to the size of the `umtx_time` struct. I don't know how I should read it at the beginning without converting it back to either one.

RalfJung replied:
You can do `read_scalar` here and then later convert the scalar as needed.

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+        let absolute_time_flag = this.eval_libc_u32("UMTX_ABSTIME");
+
+        match op {
+            // UMTX_OP_WAIT_UINT has a private variant that enables an optimization that stops it from working across processes.
+            // Miri doesn't support that anyway, so we ignore that variant and use the same implementation for all wait ops.
```
```suggestion
            // UMTX_OP_WAIT_UINT and UMTX_OP_WAIT_UINT_PRIVATE only differ in whether they work across
            // processes or not. For Miri, we can treat them the same.
```
I was very confused at first because "private variant" sounded like this was not part of the public API. ;)

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+            // UMTX_OP_WAIT_UINT has a private variant that enables an optimization that stops it from working across processes.
+            // Miri doesn't support that anyway, so we ignore that variant and use the same implementation for all wait ops.
+            op if op == wait || op == wait_uint || op == wait_uint_private => {
+                // TODO: A better way to do this? Because I don't want to duplicate the actual logic.
```
You can do something like
```rust
let obj_layout = if op == wait { this.machine.layouts.isize } else {this.machine.layouts.u32};
```
and then write the rest of the code uniformly.

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+                        .read_scalar_atomic(&obj, AtomicReadOrd::Acquire)?
+                        .to_target_isize(this)?;
+
+                    // If obj points to something negative this can never hit, otherwise we can test.
+                    if obj_val < 0 { false } else { val == u64::try_from(obj_val).unwrap() }
```
I don't understand why negative values are different here. Is this mentioned in the docs somewhere?

I assume when the docs say "The  current  value of the variable pointed to by the obj argument is compared	with the val", they mean that the bits are compared, not the mathematical values. So probably we should just treat the `obj` as `usize` here, not `isize`.

LorrensP-2158466 replied:
because i have to handle a signed or an unsigned integer, but you're right, I can just compare at bit level, thanks!

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+                                }
+                            };
+
+                            Some((TimeoutClock::Monotonic, TimeoutAnchor::Relative, duration))
```
Please quote the relevant part of the docs for which clock and anchor you are using here.

LorrensP-2158466 replied:
I still haven't found it, even after looking at the source code. But what I found really weird is the fact that the syscall doesn't take a normal `timespec` argument into account:
```c
static int
__umtx_op_wait(struct thread *td, struct _umtx_op_args *uap,
    const struct umtx_copyops *ops)
{
	struct _umtx_time timeout, *tm_p;
	int error;

	if (uap->uaddr2 == NULL)
		tm_p = NULL;
	else {
		error = ops->copyin_umtx_time(
		    uap->uaddr2, (size_t)uap->uaddr1, &timeout);
		if (error != 0)
			return (error);
		tm_p = &timeout;
	}
	return (do_wait(td, uap->obj, uap->val, tm_p, ops->compat32, 0));
}
```
`do_wait` gets `tm_p` as the only argument which specifies a timeout. But if `uaddr2` is NULL, which is the case when you only pass in `timespec` as a timeout (in the manual), then this `tm_p` is also NULL. Which has no timeout then?

Testing this on the FReeBSD VM there is indeed no timeout?
```c
#include <sys/types.h>
#include <sys/umtx.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>

int main(){
        int futex = 123;
        int expected = 123;

        struct timespec ts;

        ts.tv_sec = 0;
        ts.tv_nsec = 200000;

        int ret = _umtx_op(&futex, UMTX_OP_WAIT, expected, &ts, NULL);

        printf("return value is: %d\n", ret);
        printf("errno: %s\n", strerror(errno));
        return 0;
}
```
This hangs forever.

What should we do here? We can't really follow the manual because it doesn't specify the Clock to use, and if we use the source code, it doesn't follow the manual.

RalfJung replied:
> DISREGARD anything said in this comment, I just now noticed that I completely misread the docs a bunch of times, I finally get it...

So what is the answer then?

RalfJung replied:
Well, put the source for that claim in a comment please. :)

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+                            return this.set_last_error_and_return(LibcError("EINVAL"), dest);
+                        }
+
+                        // Inner `timespec` must still be valid.
```
I don't understand this comment.

LorrensP-2158466 replied:
The function `read_umtx_time` returns `None` if the "inner" `timespec` struct, which is parsed via `read_timespec`, returns `None`.

I will make the comment cleared on what is happening

RalfJung replied:
That's an implementation detail of `read_umtx_time`, I don't think it even needs to be mentioned here.

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+                            TimeoutAnchor::Relative
+                        };
+
+                        Some((TimeoutClock::Monotonic, anchor, umtx_time.timeout))
```
This should respect the clock ID set in the `umtx_time`.

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+                        .futex
+                        .clone();
+
+                    // TODO: This can be cleaned up no? :)
```
Can it? Given the somewhat crazy semantics, this doesn't seem unreasonable.^^

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+                            }
+                            |ecx, unblock: UnblockKind| match unblock {
+                                UnblockKind::Ready => {
+                                    ecx.write_int(0, &dest)
```
Please quote the docs that say that we should return 0 on success.

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+                    interp_ok(())
+                } else {
+                    // The manual doesn't document what should happen if `val` is invalid, so we error out.
+                    this.set_last_error_and_return(LibcError("EINVAL"), dest)
```
This is the "inequal" case, right? that's definitely not an error, it just means another thread was faster. "EINVAL" sounds odd.
Please confirm the actual behavior with a test against the real implementation.

LorrensP-2158466 replied:
So, looking at the actual implementation, nothing really happens.

It checks if `val` equals the value read from the obj pointer. If this is equal, the thread sleeps; otherwise, the function will exit with zero.

Do you think I should still test this on a real FreeBSD system?

RalfJung replied:
So, it will return 0 both when the thread was blocked (values are equal) and when it returned immediately (values are inequal)? That is odd, usually a different return value is used for these cases. So yeah would be good to confirm on a real system.

LorrensP-2158466 replied:
I tested it out on a freebsd VM with QEMU using this [guide](https://gist.github.com/ctsrc/a1f57933a2cde9abc0f07be12889f97f) and [this](https://ftp2.uk.freebsd.org/pub/FreeBSD/releases/VM-IMAGES/14.2-RELEASE/aarch64/Latest/FreeBSD-14.2-RELEASE-arm64-aarch64.raw.xz) freebsd version.

This is the C code I used:
```c
#define _WANT_UMUTEX
#include <sys/types.h>
#include <sys/umtx.h>
#include <sys/umtx.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>

int main(){
    int futex = 123;
    int expected = 456;

    int ret = _umtx_op(&futex, UMTX_OP_WAIT, expected, NULL, NULL);
    printf("return value is: %d\n", ret);
    return 0;
}
```

This prints:
```
return value is: 0
```

This is weird, but confirms it no?

RalfJung replied:
Yeah looks like it. Odd.

Cc @m-ou-se to get another set of eyes on this (you started using these functions in https://github.com/rust-lang/rust/pull/96510). The question is, does _umtx_op really not distinguish between "comparison successful, we want to sleep and were woken up again" and "comparison failed"? It seems to always return 0 in both cases.

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+                    this.set_last_error_and_return(LibcError("EINVAL"), dest)
+                }
+            }
+            // UMTX_OP_WAKE has a private variant that enables an optimization that stops it from working across processes.
+            // Miri doesn't support that anyway, so we ignore that variant and use the same implementation for all wake ops.
```
Please adjust this comment the same way as the one above.

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+                    // must be equal to the size of that struct.
+                    let timeout = if this.ptr_is_null(uaddr2)? {
+                        if this.ptr_is_null(this.read_pointer(uaddr)?)? {
+                            // ptr is null -> no timespec
```
```suggestion
                            // Both ptrs are null -> no timespec
```

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+                };
+                let futex_ref = futex_ref.futex.clone();
+
+                let count = match val{
```
```suggestion
                let count = match val {
```
Seems like rustfmt gave up on your code.^^

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+
+                let count = match val{
+                    u64::MAX => usize::MAX, // Preserve MAX because it specifies to wake everyone.
+                    // This fits because we did `read_target_usize`.
```
This is incorrect, `usize` might be 32bit but target-usize can still be 64bit.

Also, special-casing `u64::MAX` is odd. Let's not do that.
I would suggest you do a saturating cast here, i.e. something like `usize::try_from(val).unwrap_or(usize::MAX)`.

LorrensP-2158466 replied:
This will probably also work:
```rust
let count = val.min(usize::MAX as u64) as usize;
```

But then clippy yells at me:
```
warning: casting `u64` to `usize` may truncate the value on targets with 32-bit wide pointers
   --> src/shims/unix/freebsd/sync.rs:164:29
    |
164 |                 let count = val.min(usize::MAX as u64) as usize;
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: if this is intentional allow the lint with `#[allow(clippy::cast_possible_truncation)]` ...
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cast_possible_truncation
```

but this can never happen here... I'll just stick to your solution :)

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,216 @@
+        let flags = this.read_scalar(&flags_place)?.to_u32()?;
+        let clock_id_place = this.project_field(ut, 2)?;
+        let clock_id = this.read_scalar(&clock_id_place)?.to_u32()?;
+        interp_ok(Some(UmtxTime { timeout: duration, flags, _clock_id: clock_id }))
```
Please fully interpret the struct data here:
- Instead of `flags`, have a field `abstime: bool` in the struct. `flags` must be either 0 or UMTX_ABSTIME, all other values are unsupported.
- Interpret the clock ID. This will encode whether to use the system clock or monotonic clock. The logic should probably be the same as [here](https://github.com/rust-lang/rust/blob/562b282fcc8b6ffa3770d30cd40df847ebb56182/src/tools/miri/src/shims/unix/sync.rs#L327-L339).

LorrensP-2158466 replied:
Ah, that's the clock ID. I got it!

FreeBSD has a lot of clocks:
`CLOCK_REALTIME` but also `CLOCK_REALTIME_FAST`,...
`CLOCK_MONOTONIC` but also `CLOCK_MONOTONIC_FAST`, ...
And a lot more, so I should only support the above 2, but what about its variants? Just interpret them as the default, or don't support them?

RalfJung replied:
The proper thing to do is to share the code with [here](https://github.com/rust-lang/miri/blob/9611e78d8a9ba3eae6b3741edccf2c2d63e4eb42/src/shims/time.rs#L39-L76) and have a type like
```rust
enum ClockKind { System, Monotonic }
```
and a function that converts a clock_id to that enum, and use it everywhere. Then we can centrally decide which of the variants it makes sense to support.

But I'm fine with postponing this, and just supporting CLOCK_REALTIME and CLOCK_MONOTONIC for now.

Any other, unknown clock value should throw an "unsupported" error.

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,170 @@
+
+fn wake_nobody() {
+    // TODO: _umtx_op does not return how many threads were woken up
+    // How do i test this?
```
What do you want to test?

LorrensP-2158466 replied:
the Linux and MacOS tests have something like this:
```rust
// tests/pass-dep/concurrency/linux-futex.rs
fn wake_nobody() {
    let futex = 0;

    // Wake 1 waiter. Expect zero waiters woken up, as nobody is waiting.
    unsafe {
        assert_eq!(libc::syscall(libc::SYS_futex, addr_of!(futex), libc::FUTEX_WAKE, 1), 0);
    }

    // Same, but without omitting the unused arguments.
    unsafe {
        assert_eq!(
            libc::syscall(
                libc::SYS_futex,
                addr_of!(futex),
                libc::FUTEX_WAKE,
                1,
                ptr::null::<libc::timespec>(),
                0usize,
                0,
            ),
            0,
        );
    }
}
```

But I don't know how I can really test this, given _umtx_op doesn't return a value saying how many threads were woken up.

LorrensP-2158466 replied:
Update, I used the same approach as you said in your last review comment:
> You could write a test that involves two threads blocked on the futex, and then wake up one of them and check that only one gets woken.

So I have a way I think

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,170 @@
+    // TODO: _umtx_op does not return how many threads were woken up
+    // How do i test this?
+}
+fn wake_dangling() {
```
```suggestion

fn wake_dangling() {
```

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,170 @@
+            ),
+            -1
+        );
+        // man page doesn't document but we set EINVAL for consistency
+        assert_eq!(io::Error::last_os_error().raw_os_error().unwrap(), libc::ETIMEDOUT);
```
The comment doesn't match the assert. Which one is right?

LorrensP-2158466 replied:
comment was a copy paste error, oops... :)

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,170 @@
+
+    assert!((200..1000).contains(&start.elapsed().as_millis()));
+}
+fn wait_absolute_timeout() {
```
```suggestion

fn wait_absolute_timeout() {
```
In general please take a bit more care of formatting and don't just blindly hope that rustfmt will fix everything. :)

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,170 @@
+    let futex: u32 = 123;
+
+    // Wait for 200ms, with nobody waking us up early
+    unsafe {
+        assert_eq!(
+            libc::_umtx_op(
+                ptr::from_ref(&futex).cast_mut().cast(),
+                libc::UMTX_OP_WAIT_UINT_PRIVATE,
+                123,
+                &mut libc::timespec { tv_sec: 0, tv_nsec: 200_000_000 } as *mut _ as *mut _,
+                ptr::null_mut::<libc::c_void>(),
+            ),
+            -1
+        );
```
If this tests specifically the relative timeout, please call the test `wait_relative_timeout`.

Also, the test should cover both ways of doing a relative timeout: with and without `_umtx_time`.

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,170 @@
+                libc::_umtx_op(
+                    addr_of!(FUTEX) as *mut _,
+                    libc::UMTX_OP_WAKE_PRIVATE,
+                    10, // Wake up 10 threads, but we can't check that we woken up 1.
```
You could write a test that involves two threads blocked on the futex, and then wake up one of them and check that only one gets woken.

RalfJung reviewed (commented):
Sorry for the long wait! These PRs are kinda tricky to review. Overall it looks good but I have a lot of questions. :)

When resolving these comments, please do *not* force-push or amend commits. Instead, make new commits for your new changes. We'll worry about cleaning up the git history later.

RalfJung commented:
@rustbot author
Please post `@rustbot ready` when the PR is ready for the next round of review.

LorrensP-2158466 commented:
Those are not my commits, I hope I didn't do something wrong.

As for the changes, I:
- changed the handling of `uaddr` and `uaddr2` to now being correct
- fixed/updated/added tests to correctly test behaviour of the syscall

LorrensP-2158466 commented:
@rustbot ready

RalfJung commented:
Yeah something is definitely wrong with the PR, I can't review it in this state. Please make sure only your commits are on this branch.

@rustbot author

LorrensP-2158466 commented:
I think i fixed it, only my commits are present on this branch

@rustbot ready

RalfJung commented:
In the future, please avoid rebasing unless there are conflicts. Github is a pretty bad tool for code review, and rebasing throws it off entirely, causing a bunch of extra work on my side.

RalfJung commented on src/shims/unix/freebsd/foreign_items.rs:
```diff
@@ -55,6 +56,13 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 this.write_scalar(res, dest)?;
             }
 
+            // Futex
```
```suggestion
            // Synchronization primitives
```

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,238 @@
+    /// `obj`: a pointer to the futex object (can be a lot of things, mostly *AtomicU32)
+    /// `op`: the futex operation to run
+    /// `val`: the current value of the object as a `c_long` (for wait/wake)
+    /// `uaddr`: pointer to optional parameter (mostly timeouts)
```
Given that this is now read as `target_usize`, that does not seem right?

LorrensP-2158466 replied:
I wanted to keep the doc as "global" as possible, even this doc is wrong given the entire manual. `uaddr` and `uaddr2` can be many things, pointers to structs, values, keys, .... 

Is it better to specify in the `_umtx_op` documentation that only `WAIT` and `WAKE` are implemented so the comments for the `uaddr` values can be more specific?

RalfJung replied:
Well at least the docs shouldn't say anything wrong, such as saying something is a pointer which is clearly used as an integer. :)  So maybe
```
`uaddr`: `op`-specific optional parameter, pointer-sized integer or pointer
```
and then probably the same for `uaddr2`.

LorrensP-2158466 replied:
Yeah that's fair, will do!

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,238 @@
+                    // From the manual:
+                    // The timeout is specified by passing either the address of `struct timespec`, or its
+                    // extended  variant, `struct _umtx_time`, as the `uaddr2` argument of _umtx_op().
+                    // They are distinguished by the `uaddr` value, which  must be equal
```
```suggestion
                    // They are distinguished by the `uaddr` value, which must be equal
```

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,238 @@
+                        // no timeout parameter
+                        None
+                    } else {
+                        if uaddr == umtx_time_layout.size.bytes() {
```
Please add an assert that the two structs have different sizes. Otherwise this won't work...

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,238 @@
+                                }
+                            };
+
+                            // From the FreeBSD source code:
+                            // In the function `umtx_copyin_umtx_time` in kern_umtx.c, _clock_id is set to CLOCK_REALTIME when using `timespec`
```
So the docs don't say anywhere which clock is used?

LorrensP-2158466 replied:
I haven't found anything in the docs that said to use a specific Clock ID. Only thing I did found that talked about the Real Time clock whas the operation ``UMTX_OP_CV_WAIT:
From the docs:

>Optionally, a timeout for the request may be specified.	Unlike
>other requests, the timeout value is specified  directly	 by  a
>struct  timespec,  pointed  to  by the uaddr2 argument.	If the
>CVWAIT_CLOCKID flag is provided,	the  timeout  uses  the	 clock
>from  the  c_clockid  member of the struct ucond, pointed to by
>obj argument.  Otherwise, CLOCK_REALTIME	is used, regardless of
>the  clock  identifier  possibly	 specified   in	  the	struct
>_umtx_time.   If	the CVWAIT_ABSTIME flag	is supplied, the time-
>out specifies absolute time value, otherwise it denotes a rela-
>tive time interval.

But this is not what we are doing...

RalfJung replied:
It really seems entirely undocumented. And according to https://github.com/golang/go/issues/17168#issuecomment-250235271 it actually changed, it used to use the monotonic clock but then switch to REALTIME. Wild.

Please also add a link to that github discussion, and say something like:
> FreeBSD does not seem to document which clock is used when the timeout is passed as a `timespec*`. Based on discussions online and the source code, it seems to default to CLOCK_REALTIME, so that's what we also do.

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,238 @@
+                            // RealTime clock can't be used in isolation mode.
+                            this.check_no_isolation("`_umtx_op` with `TimeoutClock::RealTime`")?;
+
+                            // `uaddr2` points to a `struct timespec`
```
```suggestion
                            // `uaddr2` points to a `struct timespec`.
```

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,238 @@
+                        None
+                    } else {
+                        if uaddr == umtx_time_layout.size.bytes() {
+                            // `uaddr2` points to a `struct _umtx_time`
```
```suggestion
                            // `uaddr2` points to a `struct _umtx_time`.
```

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,238 @@
+                                UnblockKind::Ready => {
+                                    // From the manual:
+                                    // If successful, all requests, except UMTX_SHM_CREAT and  UMTX_SHM_LOOKUP
+                                    // sub-requests  of	 the  UMTX_OP_SHM  request,  will  return  zero.
```
```suggestion
                                    // sub-requests of the  UMTX_OP_SHM  request, will return zero.
```

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,238 @@
+
+                // From the manual:
+                // If successful, all requests, except UMTX_SHM_CREAT and  UMTX_SHM_LOOKUP
+                // sub-requests  of	 the  UMTX_OP_SHM  request,  will  return  zero.
```
```suggestion
                // sub-requests of the UMTX_OP_SHM  request, will return zero.
```

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,238 @@
+                            Some((umtx_time.timeout_clock, anchor, umtx_time.timeout))
+                        } else if uaddr == timespec_layout.size.bytes() {
+                            // RealTime clock can't be used in isolation mode.
+                            this.check_no_isolation("`_umtx_op` with `TimeoutClock::RealTime`")?;
```
```suggestion
                            this.check_no_isolation("`_umtx_op` with `timespec` timeout")?;
```

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,238 @@
+
+        let timeout = if raw_id == this.eval_libc_i32("CLOCK_REALTIME") {
+            // RealTime clock can't be used in isolation mode.
+            this.check_no_isolation("`_umtx_op` with `TimeoutClock::RealTime`")?;
```
```suggestion
            this.check_no_isolation("`_umtx_op` with `CLOCK_REALTIME` timeout")?;
```

RalfJung commented on src/shims/unix/freebsd/sync.rs:
```diff
@@ -0,0 +1,238 @@
+        interp_ok(Some(UmtxTime { timeout: duration, abs_time: abs_time_flag, timeout_clock }))
+    }
+
+    /// Translate raw FreeBSD clockid to a Miri TimeoutClock
```
```suggestion
    /// Translate raw FreeBSD clockid to a Miri TimeoutClock.
    /// FIXME: share this code with the pthread and clock_gettime shims.
```

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,250 @@
+            libc::_umtx_op(
+                addr_of!(FUTEX) as *mut _,
+                libc::UMTX_OP_WAIT_UINT_PRIVATE,
+                0,
```
Is this the number of threads to wake?
What's the point of the sleep...? You disabled preemption, so this can be tested without `thread::sleep`.

LorrensP-2158466 commented:
Sorry ):, you said that a rebase would probably fix it, and it sort of did. I had never seen something like that before and could not find anything that resembled that. Hopefully it doesn't cause to much problems.

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,250 @@
+    let t2 = thread::spawn(waiter);
+
+    // Wake up 1 thread and make sure the other is still waiting
+    thread::sleep(Duration::from_millis(200));
```
Again no `sleep` needed. You should `yield_now` here once, then the two threads will be scheduled and block on the futex. Then you wake up one. There shouldn't be any `sleep` in this test.

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,250 @@
+    assert!((200..1000).contains(&start.elapsed().as_millis()));
+}
+
+fn wait_wake() {
```
```suggestion
fn wait_wake_one() {
```

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,250 @@
+    wait_wrong_val();
+    wait_relative_timeout();
+    wait_absolute_timeout();
+    wait_wake();
```
`wait_wake`  and `wait_nobody` are both testing the "wake count", right? It might make sense to test 0 as a special case, but conceptually similar tests should still go together, not be at opposite ends of the test list (and the file).

LorrensP-2158466 replied:
That indeed makes more sense. I just followed the same pattern as the tests of MacOS and Linux.

Currently, I have a `wait_wake` test function, which has 2 tests: `wake_nobody` and `wake_one`. Should I also add a `wake_multiple` test?

RalfJung replied:
> I just followed the same pattern as the tests of MacOS and Linux.

Fair, sorry for having messy existing tests.^^

> Currently, I have a wait_wake test function, which has 2 tests: wake_nobody and wake_one. Should I also add a wake_multiple test?

I'd say instead of `wake_one`, have "wake 2 out of 3". That should cover most of the interesting corner cases.

RalfJung reviewed (commented):
@rustbot author

RalfJung commented:
> Sorry ):, you said that a rebase would probably fix it, and it sort of did. I had never seen something like that before and could not find anything that resembled that. Hopefully it doesn't cause to much problems.

Yeah you must have done a  rebase or a merge first to even cause the problem that the rebase then fixed. That first step is what I referred to. :)

LorrensP-2158466 commented:
@rustbot ready

RalfJung commented:
@rustbot author

LorrensP-2158466 commented:
@rustbot ready

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,257 @@
+                );
+            }
+        });
+        thread::yield_now();
```
This means the "wake 0 threads" occurs before the main thread even goes to sleep. Doesn't that defeat the purpose of the test? The test should ensure that nobody is woken up even if there's someone waiting.

LorrensP-2158466 replied:
Yes you're right, I'll fix it.

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,257 @@
+            }
+        });
+        thread::yield_now();
+        let mut timeout = libc::timespec { tv_sec: 0, tv_nsec: 400_000_000 };
```
```suggestion
        // 10ms should be enough.
        let mut timeout = libc::timespec { tv_sec: 0, tv_nsec: 10_000_000 };
```

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,257 @@
+    }
+
+    fn wake_two_of_three() {
+        // We create 2 threads that wait on a futex with a 500ms timeout.
```
A 100ms timeout should be enough hopefully. I don't want this to become the slowest of all tests just due to all the waiting. :)

LorrensP-2158466 replied:
Yeah, fair. Half a second might be a bit extreme.

RalfJung commented on tests/pass-dep/concurrency/freebsd-futex.rs:
```diff
@@ -0,0 +1,257 @@
+        let t2 = thread::spawn(waiter);
+        let t3 = thread::spawn(waiter);
+
+        thread::yield_now();
```
```suggestion
        // Run all the waiters, so they all go to sleep.
        thread::yield_now();
```

RalfJung reviewed (commented):
Just some minor nits for the test. :)

LorrensP-2158466 commented:
The tests should be ok, at least on my machine.

@rustbot ready

RalfJung commented:
This looks great, thanks! Please squash the commits, then we can land this. Please use the `--keep-base` flag when squashing so that the force-push diff is easier to review. Then write `@rustbot ready` after you pushed the squashed PR.

@rustbot author

LorrensP-2158466 commented:
I learned something new about git today, I hope this is what you meant :).

@rustbot ready

RalfJung commented:
Looks good, thanks :)
